resources/nersc_supervisor.sh

```diff
@@ -0,0 +1,248 @@
+ 
+  if [[ ${NS_LOG_DETAIL} -ge ${level} ]]
+    then
+      echo "$(date) $2" >> "${NS_LOG}"
+    fi
+  return 0
+ }
+
+###
+# Return info about slurm jobs
+function getrunning {
+  # no arguments
+  logit 2 "getrunning"
+  if ! rawinfo=$(${SQUEUE} -h -o "%.18i %.25j %.2t %.10M %.42k %R" -u icecubed -q xfer -M escori)
+    then
+      logit 0 "SLURM is not working"
+      return 1
+    fi
+  echo "${rawinfo}"
+  return 0
+}
+
+###
+# Launch a process of the appropriate type
+function launch {
+  # argument:  class of job to submit:  e.g. nersc_mover
+  # I expect a shell script which in turn invokes a python script
+  #  but that can be changed at will
+  if [[ ! -f ${NS_SCRIPT_PATH}/$1".sh" ]]
+    then
+      logit 0 "Submit script $1.sh was not found"
+      return 1
+    fi
+  logit 2 "launch $1"
+  if ! ${SBATCH}  -o "${NS_SLURM_LOG}/slurm-$1-%j.out" -q xfer -M escori -t 12:00:00 "${NS_SCRIPT_PATH}/$1.sh"
+    then
+      logit 0 "Submit of $1 failed"
+      return 1
+    fi
+  return 0
+}
+
+###
+# Check the slurm job activity information, count the incomplete entries
+
+function countIncomplete {
+  # No arguments
+  #
+  logit 2 'Entering countIncomplete'
+  twoweeksbefore=$(/usr/bin/date +%s | /usr/bin/awk '{printf("%s\n",strftime("%Y-%m-%d",$1));}')
+  # NOTA BENE:  The same job number can appear several times in the output.
+  # Therefore I have to sort -u the result.
+  # bash will treat the contents of "listing" as a single line, of course.
+  listing=$(${SACCT} -n -b -S "${twoweeksbefore}" -M escori -p | /usr/bin/awk '{split($0,b,"|");split(b[1],a,".");if(b[2]!="COMPLETED")print a[1],b[2],b[3],"|";}' |/usr/bin/sort -u)
+  if [[ "${listing}" == "" ]]
+    then
+        echo 0
+        return 0
+    fi
+  # Steve doesn't think I need to log the incomplete info
+  #logit 2 "${listing}"
+  # Count the number of not-completed, skipping the terminal "|" if any
+  echo "${listing}" | /usr/bin/awk '{n=split($0,a,"|");m=n-1;if(m<0)m=0;print m;}'
+  return 0
+}
+
+###################################################################
+# This is where the main work really begins.
+#############################################
+#
+
+# First find and set the logging level
+
+if [[ "$1" == "" ]]
+  then
+    NS_LOG_DETAIL=0
+  else
+    NS_LOG_DETAIL=$1
+  fi
+# Log (or not, depending on the level)
+logit 1 "Invocation"
+
+# Not used yet
+
+
+
+
+##################
+# Start the work #
+##################
+
+#if ! incomplete=$(countIncomplete)
+#  then
+#    exit 1
+#  fi
+
+###
+# How many slurm jobs are running right now?
+###
+if ! whatWeHave=$(getrunning)
+  then
+    exit 2
+  fi
+
+###
+# Loop over the list of expected job types
+# At the moment this info is hard-wired in NS_DESIRED, but
+# we should load from a json initialization file instead
+# Keeping track of activeJobs is future-proofing the system
+# I should count them all and compare with maxjobs before
+# submitting anything.
+###
+activeJobs=0
```

Comment:
It looks like the code below is counting up the jobs from those specified in `NS_DESIRED`.
However, if there were other jobs running (of a kind not specified in `NS_DESIRED`) it looks like it wouldn't be counted as part of `activeJobs`.
Since `activeJobs` is just future-proofing, there isn't really anything to change here yet.
Just something (non-specified jobs) to keep in mind when building that future proofing at a later date.

Reply:
Yes, I'll have to have a full list of what's expected--because slurm jobs may not be HPSS jobs.